File: packages/sdk-java/src/main/java/org/melaya/TradeAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Live trading API — credentialed order placement, account state, and
 * position management on a CONNECTED exchange.
 *
 * <p>Every method POSTs to {@code https://api.melaya.org/api/v1/private/<op>}; the server
 * resolves your connected exchange credential (referenced by {@code apiKeyId} — see
 * {@link AccountAPI#keys}) and forwards the call to the venue through Melaya's
 * in-house Rust engine. Responses share an envelope:
 * {@code {ok, exchange, operation, orderId, clientOrderId, payload, data, ...}}.
 *
 * <p><strong>WARNING:</strong> the write methods (createOrder, cancelOrder, amendOrder,
 * cancelAllOrders, cancelPlanOrders, closePosition, setLeverage, setMarginMode,
 * setPositionMode) hit the REAL venue with REAL funds. For risk-free testing use
 * {@link SimAPI} (paper) or a paper strategy instead.
 */
public class TradeAPI {

    private final HttpClient http;

    TradeAPI(HttpClient http) {
        this.http = http;
    }

    private JsonNode op(String op, Map<String, Object> body) {
        return http.post("/api/v1/private/" + op, body);
    }

    /** Build a nullable-omitting body map from alternating key/value pairs. */
    private static Map<String, Object> body(Object... kv) {
        if (kv.length % 2 != 0) throw new IllegalArgumentException("body() requires even number of args");
        Map<String, Object> m = new LinkedHashMap<>();
        for (int i = 0; i < kv.length; i += 2) {
            if (kv[i + 1] != null) m.put(String.valueOf(kv[i]), kv[i + 1]);
        }
        return m;
    }

    // ── Account state (reads) ─────────────────────────────────────────────────

    /** Live account balance on a connected venue. */
    public JsonNode balance(String exchange, String apiKeyId, String keyId, String marketType) {
        return op("balance", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "keyId", keyId,
                "marketType", marketType));
    }

    /** Live open positions. */
    public JsonNode positions(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("positions", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Historical positions (venue-dependent). */
    public JsonNode positionsHistory(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("positions-history", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Resting (open) orders. */
    public JsonNode openOrders(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("open-orders", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** All orders (open + recent). */
    public JsonNode orders(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("orders", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Closed/filled orders. */
    public JsonNode closedOrders(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("closed-orders", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Your trade (fill) history. */
    public JsonNode myTrades(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("my-trades", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Extended trade history (venue-dependent). */
    public JsonNode myTradesHistory(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("my-trades-history", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Resting conditional/plan (trigger) orders. */
    public JsonNode planOrders(String exchange, String apiKeyId, String marketType, String symbol) {
        return op("plan-orders", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "marketType", marketType,
                "symbol", symbol));
    }

    /** Current leverage for a symbol. */
    public JsonNode leverage(String exchange, String apiKeyId, String symbol, String marketType) {
        return op("leverage", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "symbol", symbol,
                "marketType", marketType));
    }

    /** Leverage tiers / brackets for a symbol. */
    public JsonNode leverageTiers(String exchange, String apiKeyId, String symbol, String marketType) {
        return op("leverage-tiers", body(
                "exchange", exchange,
                "apiKeyId", apiKeyId,
                "symbol", symbol,
                "marketType", marketType));
    }

    // ── Order placement & management (LIVE writes — real funds) ───────────────

    /**
     * Place a live order on the venue. <strong>WARNING: real money.</strong>
     * {@code stopPrice}, {@code takeProfitPrice}, and {@code reduceOnly} are
     * folded into the {@code params} sub-object as required by the contract.
     */
    public JsonNode createOrder(String exchange, String apiKeyId, String symbol,
                                String side, double amount, String type, Double price,
                                String marketType, Double leverage, String clientOrderId,
                                Double stopPrice, Double takeProfitPrice, Boolean reduceOnly,
                                Map<String, Object> params) {
        Map<String, Object> p = new LinkedHashMap<>();
        if (params != null) p.putAll(params);
        if (stopPrice != null)       p.put("stopPrice",       stopPrice);
        if (takeProfitPrice != null) p.put("takeProfitPrice", takeProfitPrice);
        if (reduceOnly != null)      p.put("reduceOnly",      reduceOnly);

        Map<String, Object> b = body(
                "exchange",       exchange,
                "apiKeyId",       apiKeyId,
                "symbol",         symbol,
                "side",           side,
                "amount",         amount,
                "type",           type,
                "price",          price,
                "marketType",     marketType,
                "leverage",       leverage,
                "clientOrderId",  clientOrderId);
        if (!p.isEmpty()) b.put("params", p);
        return op("create-order", b);
    }

    /** Cancel a live order by id. <strong>WARNING.</strong> */
    public JsonNode cancelOrder(String exchange, String apiKeyId, String orderId,
                                String clientOrderId, String symbol, String marketType) {
        return op("cancel-order", body(
                "exchange",       exchange,
                "apiKeyId",       apiKeyId,
                "orderId",        orderId,
                "clientOrderId",  clientOrderId,
                "symbol",         symbol,
                "marketType",     marketType));
    }

    /** Amend (modify) a live order. <strong>WARNING.</strong> */
    public JsonNode amendOrder(String exchange, String apiKeyId, String orderId,
                               String symbol, Double amount, Double price) {
        return op("amend-order", body(
                "exchange",  exchange,
                "apiKeyId",  apiKeyId,
                "orderId",   orderId,
                "symbol",    symbol,
                "amount",    amount,
                "price",     price));
    }

    /** Cancel every open order (optionally scoped to a symbol). <strong>WARNING.</strong> */
    public JsonNode cancelAllOrders(String exchange, String apiKeyId, String symbol, String marketType) {
        return op("cancel-all-orders", body(
                "exchange",    exchange,
                "apiKeyId",    apiKeyId,
                "symbol",      symbol,
                "marketType",  marketType));
    }

    /** Cancel resting plan/trigger orders. <strong>WARNING.</strong> */
    public JsonNode cancelPlanOrders(String exchange, String apiKeyId, String symbol, String marketType) {
        return op("cancel-plan-orders", body(
                "exchange",    exchange,
                "apiKeyId",    apiKeyId,
                "symbol",      symbol,
                "marketType",  marketType));
    }

    /** Close an open position (market reduce-only). <strong>WARNING.</strong> */
    public JsonNode closePosition(String exchange, String apiKeyId, String symbol, String marketType) {
        return op("close-position", body(
                "exchange",    exchange,
                "apiKeyId",    apiKeyId,
                "symbol",      symbol,
                "marketType",  marketType));
    }

    /** Set leverage for a symbol. <strong>WARNING.</strong> */
    public JsonNode setLeverage(String exchange, String apiKeyId, String symbol,
                                double leverage, String marketType) {
        return op("set-leverage", body(
                "exchange",    exchange,
                "apiKeyId",    apiKeyId,
                "symbol",      symbol,
                "leverage",    leverage,
                "marketType",  marketType));
    }

    /** Set margin mode (cross/isolated). <strong>WARNING.</strong> */
    public JsonNode setMarginMode(String exchange, String apiKeyId, String marginMode,
                                  String symbol, String marketType) {
        return op("set-margin-mode", body(
                "exchange",    exchange,
                "apiKeyId",    apiKeyId,
                "marginMode",  marginMode,
                "symbol",      symbol,
                "marketType",  marketType));
    }

    /** Set position mode (one-way / hedge). <strong>WARNING.</strong> */
    public JsonNode setPositionMode(String exchange, String apiKeyId, Boolean hedged,
                                    String mode, String marketType) {
        return op("set-position-mode", body(
                "exchange",    exchange,
                "apiKeyId",    apiKeyId,
                "hedged",      hedged,
                "mode",        mode,
                "marketType",  marketType));
    }
}

File: packages/sdk-kotlin/src/main/kotlin/org/melaya/TradeAPI.kt
package org.melaya

import org.json.JSONObject

/**
 * Live trading API — credentialed order placement, account state, and
 * position management on a CONNECTED exchange.
 *
 * Every method POSTs to `https://api.melaya.org/api/v1/private/<op>`; the server
 * resolves your connected exchange credential (referenced by `apiKeyId` — see
 * `account.keys()`) and forwards the call to the venue through Melaya's in-house
 * Rust engine. Responses share an envelope:
 * `{ok, exchange, operation, orderId, clientOrderId, payload, data, ...}`.
 *
 * **WARNING:** the write methods (createOrder, cancelOrder, amendOrder,
 * cancelAllOrders, cancelPlanOrders, closePosition, setLeverage, setMarginMode,
 * setPositionMode) hit the REAL venue with REAL funds. For risk-free testing use
 * [SimAPI] (paper) or a paper strategy instead.
 */
class TradeAPI internal constructor(private val http: HttpClient) {

    private fun op(op: String, body: Map<String, Any?>): JSONObject =
        http.post("/api/v1/private/$op", body).asObject()

    /** Build a nullable-omitting map. */
    private fun b(vararg kv: Pair<String, Any?>): Map<String, Any?> =
        kv.filter { it.second != null }.toMap()

    // ── Account state (reads) ─────────────────────────────────────────────────

    /** Live account balance on a connected venue. */
    fun balance(
        exchange: String,
        apiKeyId: String? = null,
        keyId: String? = null,
        marketType: String? = null,
    ): JSONObject = op("balance", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "keyId"      to keyId,
        "marketType" to marketType,
    ))

    /** Live open positions. */
    fun positions(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("positions", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Historical positions (venue-dependent). */
    fun positionsHistory(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("positions-history", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Resting (open) orders. */
    fun openOrders(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("open-orders", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** All orders (open + recent). */
    fun orders(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("orders", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Closed/filled orders. */
    fun closedOrders(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("closed-orders", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Your trade (fill) history. */
    fun myTrades(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("my-trades", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Extended trade history (venue-dependent). */
    fun myTradesHistory(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("my-trades-history", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Resting conditional/plan (trigger) orders. */
    fun planOrders(
        exchange: String,
        apiKeyId: String? = null,
        marketType: String? = null,
        symbol: String? = null,
    ): JSONObject = op("plan-orders", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marketType" to marketType,
        "symbol"     to symbol,
    ))

    /** Current leverage for a symbol. */
    fun leverage(
        exchange: String,
        apiKeyId: String? = null,
        symbol: String? = null,
        marketType: String? = null,
    ): JSONObject = op("leverage", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "symbol"     to symbol,
        "marketType" to marketType,
    ))

    /** Leverage tiers / brackets for a symbol. */
    fun leverageTiers(
        exchange: String,
        apiKeyId: String? = null,
        symbol: String? = null,
        marketType: String? = null,
    ): JSONObject = op("leverage-tiers", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "symbol"     to symbol,
        "marketType" to marketType,
    ))

    // ── Order placement & management (LIVE writes — real funds) ───────────────

    /**
     * Place a live order on the venue. **WARNING: real money.**
     * [stopPrice], [takeProfitPrice], and [reduceOnly] are folded into the
     * `params` sub-object as required by the contract.
     */
    fun createOrder(
        exchange: String,
        symbol: String,
        side: String,
        amount: Double,
        apiKeyId: String? = null,
        type: String = "market",
        price: Double? = null,
        marketType: String? = null,
        leverage: Double? = null,
        clientOrderId: String? = null,
        stopPrice: Double? = null,
        takeProfitPrice: Double? = null,
        reduceOnly: Boolean? = null,
        params: Map<String, Any?> = emptyMap(),
    ): JSONObject {
        val p = buildMap {
            putAll(params)
            if (stopPrice != null)       put("stopPrice",       stopPrice)
            if (takeProfitPrice != null) put("takeProfitPrice", takeProfitPrice)
            if (reduceOnly != null)      put("reduceOnly",      reduceOnly)
        }
        val body = buildMap {
            put("exchange", exchange)
            if (apiKeyId != null)      put("apiKeyId",       apiKeyId)
            put("symbol", symbol)
            put("side", side)
            put("amount", amount)
            put("type", type)
            if (price != null)         put("price",          price)
            if (marketType != null)    put("marketType",     marketType)
            if (leverage != null)      put("leverage",       leverage)
            if (clientOrderId != null) put("clientOrderId",  clientOrderId)
            if (p.isNotEmpty())        put("params",         p)
        }
        return http.post("/api/v1/private/create-order", body).asObject()
    }

    /** Cancel a live order by id. **WARNING.** */
    fun cancelOrder(
        exchange: String,
        apiKeyId: String? = null,
        orderId: String? = null,
        clientOrderId: String? = null,
        symbol: String? = null,
        marketType: String? = null,
    ): JSONObject = op("cancel-order", b(
        "exchange"      to exchange,
        "apiKeyId"      to apiKeyId,
        "orderId"       to orderId,
        "clientOrderId" to clientOrderId,
        "symbol"        to symbol,
        "marketType"    to marketType,
    ))

    /** Amend (modify) a live order. **WARNING.** */
    fun amendOrder(
        exchange: String,
        apiKeyId: String? = null,
        orderId: String? = null,
        symbol: String? = null,
        amount: Double? = null,
        price: Double? = null,
    ): JSONObject = op("amend-order", b(
        "exchange"  to exchange,
        "apiKeyId"  to apiKeyId,
        "orderId"   to orderId,
        "symbol"    to symbol,
        "amount"    to amount,
        "price"     to price,
    ))

    /** Cancel every open order (optionally scoped to a symbol). **WARNING.** */
    fun cancelAllOrders(
        exchange: String,
        apiKeyId: String? = null,
        symbol: String? = null,
        marketType: String? = null,
    ): JSONObject = op("cancel-all-orders", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "symbol"     to symbol,
        "marketType" to marketType,
    ))

    /** Cancel resting plan/trigger orders. **WARNING.** */
    fun cancelPlanOrders(
        exchange: String,
        apiKeyId: String? = null,
        symbol: String? = null,
        marketType: String? = null,
    ): JSONObject = op("cancel-plan-orders", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "symbol"     to symbol,
        "marketType" to marketType,
    ))

    /** Close an open position (market reduce-only). **WARNING.** */
    fun closePosition(
        exchange: String,
        symbol: String,
        apiKeyId: String? = null,
        marketType: String? = null,
    ): JSONObject = op("close-position", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "symbol"     to symbol,
        "marketType" to marketType,
    ))

    /** Set leverage for a symbol. **WARNING.** */
    fun setLeverage(
        exchange: String,
        symbol: String,
        leverage: Double,
        apiKeyId: String? = null,
        marketType: String? = null,
    ): JSONObject = op("set-leverage", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "symbol"     to symbol,
        "leverage"   to leverage,
        "marketType" to marketType,
    ))

    /** Set margin mode (cross/isolated). **WARNING.** */
    fun setMarginMode(
        exchange: String,
        marginMode: String,
        apiKeyId: String? = null,
        symbol: String? = null,
        marketType: String? = null,
    ): JSONObject = op("set-margin-mode", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "marginMode" to marginMode,
        "symbol"     to symbol,
        "marketType" to marketType,
    ))

    /** Set position mode (one-way / hedge). **WARNING.** */
    fun setPositionMode(
        exchange: String,
        apiKeyId: String? = null,
        hedged: Boolean? = null,
        mode: String? = null,
        marketType: String? = null,
    ): JSONObject = op("set-position-mode", b(
        "exchange"   to exchange,
        "apiKeyId"   to apiKeyId,
        "hedged"     to hedged,
        "mode"       to mode,
        "marketType" to marketType,
    ))
}
